### PR TITLE
aea publish --push-missing

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

aea publish --push-missing

## Fixes

https://github.com/fetchai/agents-aea/issues/2531

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/cli/publish.py [hunk 1]
```diff
@@ -37,36 +40,46 @@
...
     CONNECTIONS,
     CONTRACTS,
     DEFAULT_AEA_CONFIG_FILE,
+    ITEM_TYPE_PLURAL_TO_TYPE,
     PROTOCOLS,
     SKILLS,
 )
 
 
+PUSH_ITEMS_FLAG = "--push-missing"
```
Comment (DavidMinarsch) [inline aea/cli/publish.py:56]: not actually used?

File: aea/cli/publish.py [hunk 2]
```diff
@@ -112,7 +125,7 @@ def _check_is_item_in_registry_mixed(
             click.echo("Found!")
         except click.ClickException as e:
             raise click.ClickException(
-                f"Package not found neither in local nor in remote registry: {str(e)}"
+                f"Package not found neither in local nor in remote registry: {str(e)}. You can try to add {PUSH_ITEMS_FLAG}."
```
Comment (DavidMinarsch) [inline aea/cli/publish.py:128]: Does it make sense?

File: aea/cli/publish.py [hunk 3]
```diff
@@ -139,33 +152,185 @@ def _check_is_item_in_local_registry(
...
-            if is_mixed:
-                _check_is_item_in_registry_mixed(
-                    PublicId.from_str(str(public_id)), item_type_plural, registry_path,
-                )
-            else:
-                _check_is_item_in_local_registry(
-                    PublicId.from_str(str(public_id)), item_type_plural, registry_path,
-                )
+
+    registry = LocalRegistry(ctx, is_mixed)
```
Comment (DavidMinarsch) [inline aea/cli/publish.py:331]: why is_mixed?

File: aea/cli/publish.py [hunk 4]
```diff
@@ -139,33 +152,185 @@ def _check_is_item_in_local_registry(
         )
     except click.ClickException as e:
         raise click.ClickException(
-            f"Dependency is missing. {str(e)}\nPlease push it first and then retry."
+            f"Dependency is missing. {str(e)}\nPlease push it first and then retry or use {PUSH_ITEMS_FLAG} flag to push automatically"
         )
 
 
-def _save_agent_locally(ctx: Context, is_mixed: bool = False) -> None:
+class BaseRegistry(ABC):
```
Comment (marcofavorito) [inline aea/cli/publish.py:159]: this is very useful to have. Maybe move to another Python module like `aea.cli.registry.base`. OR, also consider including it into AEA core, e.g. `aea.registry` (note: `aea.registries.*` is just for internal use, not related to AEA registry)

Comment (marcofavorito) [reply aea/cli/publish.py:159]: The reason is, in future refactorings, we surely want to use this class for registry-related features.

File: aea/cli/publish.py [hunk 5]
```diff
@@ -139,33 +152,185 @@ def _check_is_item_in_local_registry(
...
+        :return: None
+        """
+
+        try:
+            _check_is_item_in_remote_registry(public_id, item_type_plural)
+        except click.ClickException as e:
+            raise click.ClickException(
+                f"Package not found in remote registry: {str(e)}. You can try to add {PUSH_ITEMS_FLAG} flag."
+            )
+
```
Comment (DavidMinarsch) [inline aea/cli/publish.py:288]: Some code duplication with above? Remove the old one?

File: aea/cli/publish.py (hunk 4 above)
Comment (marcofavorito) [reply aea/cli/publish.py:159]: the same applies below for implementations of this interface

File: aea/cli/publish.py [hunk 6]
```diff
@@ -139,33 +152,185 @@ def _check_is_item_in_local_registry(
...
+        try:
+            self.check_item_present(item_type_plural, public_id)
+        except Exception as e:
+            raise click.ClickException(
+                f"Failed to find item after push: {item_type_plural} {public_id}: {e}"
+            ) from e
+
+
+class LocalRegistry(BaseRegistry):
```
Comment (DavidMinarsch) [inline aea/cli/publish.py:218]: Might as well have a mixed case?

Comment (marcofavorito) [reply aea/cli/publish.py:218]: In my opinion, the mixed case should be handled by a separate registry class, which is a composition of `LocalRegistry` and `RemoteRegistry`.

Comment (marcofavorito) [reply aea/cli/publish.py:218]: But for the scope of this PR it is probably OK to have the mixed mode implemented in `LocalRegistry`.

File: aea/cli/publish.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/cli/publish.py:159]: Yes, although please keep it in CLI for now; let's avoid doing large changes during minors.

**Review by marcofavorito — APPROVED** [review]

File: aea/cli/publish.py (hunk 1 above)
Comment (solarw) [reply aea/cli/publish.py:56]: used in messages

File: aea/cli/publish.py (hunk 2 above)
Comment (solarw) [reply aea/cli/publish.py:128]: yes, just a hint for the user to add a flag to publish automatically

File: aea/cli/publish.py (hunk 6 above)
Comment (solarw) [reply aea/cli/publish.py:218]: updated

File: aea/cli/publish.py (hunk 3 above)
Comment (solarw) [reply aea/cli/publish.py:331]: updated

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**